src/compiler/parser.ts

```diff
@@ -2818,7 +2821,23 @@ module ts {
                     break;
                 }
 
-                leftOperand = makeBinaryExpression(leftOperand, parseTokenNode(), parseBinaryExpressionOrHigher(newPrecedence));
+                if (token === SyntaxKind.AsKeyword) {
+                    // Make sure we *do* perform ASI for constructs like this:
+                    //    var x = foo
+                    //    as (Bar)
+                    // This should be parsed as an initialized variable, followed
+                    // by a function call to 'as' with the argument 'Bar'
+                    if (canParseSemicolon()) {

```

Comment:
Can you just check scanner.hasPrecedingLineBreak instead? I don't really like canParseSemicolon that much.